crates/rattler_conda_types/src/version_spec/version_tree.rs

```diff
@@ -198,4 +228,158 @@ mod tests {
+        );
+        assert_eq!(recognize_version::<Err>("3."), Ok((".", "3")));
+        assert_eq!(recognize_version::<Err>("3.*"), Ok((".*", "3")));
+
+        let versions = [
+            // Implicit epoch of 0
+            "1.0a1",
+            "1.0a2.dev456",
+            "1.0a12.dev456",
+            "1.0a12",
+            "1.0b1.dev456",
+            "1.0b2",
+            "1.0b2.post345.dev456",
+            "1.0b2.post345",
+            "1.0c1.dev456",
+            "1.0c1",
+            "1.0c3",
+            "1.0rc2",
+            "1.0.dev456",
+            "1.0",
+            "1.0.post456.dev34",
+            "1.0.post456",
+            "1.1.dev1",
+            "1.2.r32+123456",
+            "1.2.rev33+123456",
+            "1.2+abc",
+            "1.2+abc123def",
+            "1.2+abc123",
+            "1.2+123abc",
+            "1.2+123abc456",
+            "1.2+1234.abc",
+            "1.2+123456",
+            // Explicit epoch of 1
+            "1!1.0a1",
+            "1!1.0a2.dev456",
+            "1!1.0a12.dev456",
+            "1!1.0a12",
+            "1!1.0b1.dev456",
+            "1!1.0b2",
+            "1!1.0b2.post345.dev456",
+            "1!1.0b2.post345",
+            "1!1.0c1.dev456",
+            "1!1.0c1",
+            "1!1.0c3",
+            "1!1.0rc2",
+            "1!1.0.dev456",
+            "1!1.0",
+            "1!1.0.post456.dev34",
+            "1!1.0.post456",
+            "1!1.1.dev1",
+            "1!1.2.r32+123456",
+            "1!1.2.rev33+123456",
+            "1!1.2+abc",
+            "1!1.2+abc123def",
+            "1!1.2+abc123",
+            "1!1.2+123abc",
+            "1!1.2+123abc456",
+            "1!1.2+1234.abc",
+            "1!1.2+123456",
+        ];
+
+        for version_str in versions {
+            assert_eq!(recognize_version::<Err>(version_str), Ok(("", version_str)));
+        }
+    }
+
+    #[test]
+    fn test_parse_version_epoch() {
+        type Err<'a> = nom::error::Error<&'a str>;
+
+        assert_eq!(
+            parse_version_epoch::<Err>("1!1.0b2.post345.dev456"),
+            Ok(("1.0b2.post345.dev456", 1))
+        );
+    }
+
+    #[test]
+    fn test_recognize_constraint() {
+        type Err<'a> = nom::error::Error<&'a str>;
+
+        assert_eq!(recognize_constraint::<Err>("*"), Ok(("", "*")));
+        assert_eq!(recognize_constraint::<Err>("3.8"), Ok(("", "3.8")));
+        assert_eq!(recognize_constraint::<Err>("3.8*"), Ok(("", "3.8*")));
+        assert_eq!(recognize_constraint::<Err>("3.8.*"), Ok(("", "3.8.*")));
+        assert_eq!(recognize_constraint::<Err>(">=3.8.*"), Ok(("", ">=3.8.*")));
+        assert_eq!(
+            recognize_constraint::<Err>(">=3.8.*<3.9"),
+            Ok(("<3.9", ">=3.8.*"))
+        );
+        assert_eq!(
+            recognize_constraint::<Err>(">=3.8.*,<3.9"),
+            Ok((",<3.9", ">=3.8.*"))
+        );
+    }
+
+    #[test]
+    fn issue_204() {
+        assert!(VersionTree::try_from(">=3.8<3.9").is_err());
```

Comment:
Maybe also do a test for a invalid epoch? E.g. `12.23!123.321` shouldn't work (afaik epochs can only be simple integers).

Reply:
Done!